File: algorithms_c/sort/bubble_rec.c
/* Source: ://github.com/TheAlgorithms/C/blob/master/sorting/bubble_rec_sort_recursion.c */
#include <stdbool.h>
#include <stdlib.h>
#include "bubble_rec.h"
static void swap(long *first, long *second)
{
    long temp = *first;
    *first = *second;
    *second = temp;
}

/**
 * Bubble sort algorithm implements using recursion
 * @param arr array to be sorted
 * @param size size of array
 */
void bubble_rec_sort(long *arr, long size)
{
    if (size == 1)
    {
        return;
    }
    bool swapped = false;
    for (int i = 0; i < size - 1; ++i)
    {
        if (arr[i] > arr[i + 1])
        {
            swap(arr + i, arr + i + 1);
            swapped = true;
        }
    }
    if (swapped)
    {
        bubble_rec_sort(arr, size - 1);
    }
}

File: algorithms_java/sort/BubbleRec.java
/* Source: https://github.com/TheAlgorithms/Java/blob/master/Sorts/BubbleSortRecursion.java */
public class BubbleRec{
	public static void bubbleSort(Long A[], int n, int i, int j){
		Long swap = new Long(0);
		if(i >= n-1){
			return;
		}	
		if(j < n - 1){
			bubbleSort(A, n, i, j + 1);
			if(A[j] > A[j+1]){
				swap = A[j];
				A[j] = A[j+1];
				A[j+1] = swap;
				swap = null;
			}
		}
		if(j == 0)
			bubbleSort(A, n, i +1, 0);
	}
}
